src/firefly/js/ui/panel/CollapsiblePanel.jsx

```diff
@@ -25,85 +24,64 @@ export const CollapseHeaderCorner = {
                             BottomLeft: 0b1000
                         };
 
-function collapsibleStateFromProps(props) {
-    // component key needs to be defined if the state needs to be saved though unmount/mount
-    let isOpen = props.componentKey && getComponentState(props.componentKey).isOpen;
-    if (!isBoolean(isOpen)) {
-        // use property to initialize state
-        isOpen = props.isOpen? props.isOpen: false;
-    }
 
-    return {isOpen};
-}
 
-export class CollapsiblePanel extends PureComponent {
+export function CollapsiblePanel({ componentKey, header, isOpen=false, headerRoundCorner=0, headerStyle, contentStyle,
+                                     style, borderStyle=CollapseBorder.Oneborder, onToggle, children }) {
+    function getIsOpenFromComponentState() {
+        if (!componentKey) return isOpen;
+        return getComponentState(componentKey)?.isOpen ?? isOpen;
+    }
+    const isOpenGlobalState = getIsOpenFromComponentState();
 
-    constructor(props) {
-        super(props);
+    const [isOpenState, setIsOpenState] = useState(isOpenGlobalState);
 
-        this.state = collapsibleStateFromProps(props);
+    // if global state has updated, update local state
+    if (isOpenGlobalState !== isOpenState) setIsOpenState(isOpenGlobalState);
 
-        this.handleClick = this.handleClick.bind(this);
-        this.getContentHeight = this.getContentHeight.bind(this);
-    }
+    function handleClick() {
+        const newIsOpen = !isOpenState;
 
-    static getDerivedStateFromProps(props) {
-        return collapsibleStateFromProps(props);
+        if (componentKey) { // update global state
+            dispatchComponentStateChange(componentKey, {isOpen: newIsOpen});
+        }
+        setIsOpenState(newIsOpen);// update local state
+        onToggle?.(newIsOpen);
     }
 
-    handleClick() {
-        const {componentKey, onToggle} = this.props;
-        const isOpen = !this.state.isOpen;
 
-        if (componentKey) {
-            dispatchComponentStateChange(componentKey, {isOpen});
-        }
-        this.setState({isOpen});
+    const contentBorderClassName = ['', ' CollapsiblePanel__Content-oneborder',
+        ' CollapsiblePanel__Content-twoborder', ' CollapsiblePanel__Content-threeborder'];
+    let headerClassName = 'CollapsiblePanel__Header';
+    let contentClassName = 'CollapsiblePanel__Content';
+    let headerCorner = '';
+    ['TopLeft', 'TopRight', 'BottomRight', 'BottomLeft'].forEach((corner) => {
+        headerCorner += (CollapseHeaderCorner[corner]&headerRoundCorner ? ' 0.5em' : ' 0em');
+    });
```

Comment:
I had a comment on this `forEach` last time but I must have forgot to push the add comment button.

This is a classic `reduce` and using `reduce` makes the code more clean. Also an array defined and the operated on also is confusing. Put the array outside of the component at the top around line 26.
`const corners= ['TopLeft', 'TopRight', 'BottomRight', 'BottomLeft']`
Then...
```js
 const headerCorner= corners.reduce((str,corner) => {
            str += (CollapseHeaderCorner[corner]&headerRoundCorner ? ' 0.5em' : ' 0em');
            return str;
        },'');
```

Reply:
I agree this is more clean.